File: built-in plugins dsp/mb-time-stretch.cpp
/**
 * MB Time Stretcher
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : High-quality time stretching without pitch change
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_TIME_STRETCH_H
#define MB_TIME_STRETCH_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbTimeStretch : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-time-stretch";
    static constexpr const char* PLUGIN_NAME    = "MB Time Stretcher";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float ratio = 1f;  // range [0.25, 4]
    float quality = 2f;  // range [0, 3]
    float preserveTransients = 1f;  // range [0, 1]
    };

    MbTimeStretch() = default;
    ~MbTimeStretch() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.ratio = std::clamp(params.ratio, 0.25f, 4f);
        params.quality = std::clamp(params.quality, 0f, 3f);
        params.preserveTransients = std::clamp(params.preserveTransients, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Time Stretcher
        return input;
    }
};

#endif // MB_TIME_STRETCH_H

File: built-in plugins dsp/mb-time-stretch.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbTimeStretchPlugin: PluginDefinition = {
  id: "mb-time-stretch",
  slug: "mb-time-stretch",
  name: "MB Time Stretcher",
  category: "effect",
  type: "delay" as any,
  version: "1.0.0",
  description: "High-quality time stretching without pitch change",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "ratio",
      name: "Time Ratio",
      type: "float",
      defaultValue: 1,
      minValue: 0.25,
      maxValue: 4,
      automatable: true,
    },
    {
      id: "quality",
      name: "Quality",
      type: "float",
      defaultValue: 2,
      minValue: 0,
      maxValue: 3,
      automatable: false,
    },
    {
      id: "preserveTransients",
      name: "Preserve Transients",
      type: "float",
      defaultValue: 1,
      minValue: 0,
      maxValue: 1,
      automatable: false,
    },
  ],
  defaultPreset: { ratio: 1, quality: 2, preserveTransients: 1 },
};

export default MbTimeStretchPlugin;
